app: Add custom AT command to support nRF Cloud provisioning service (#258)
nrfconnect/ncs-serial-modem: doc/app/at_provisioning.rst

.. _SM_AT_PROVISIONING:

nRF Provisioning AT commands
*****************************

.. contents::
   :local:
   :depth: 2

.. note::

   These AT commands are `Experimental <Software maturity levels_>`_.

This page describes AT commands for the nRF Device Provisioning service.
The provisioning client connects to the `nRF Cloud Provisioning Service`_ over CoAP/DTLS

> and applies device configuration commands (credentials, settings, firmware updates) issued from the cloud.

Reviewer: ```suggestion
The provisioning client connects to the `nRF Cloud Provisioning Service`_ over CoAP or DTLS and applies device configuration commands (credentials, settings, and firmware updates) issued from the cloud.
```

> 

Reviewer 2: A link to provisioning service is above but...
That contains very long piece of information making it more complex than required to make a trial. Would it make sense to add here a 3-5 sentence paragraph on how to test this with 9151DK, which I assume would have certificates in place. I'll suggest something. May be a bit clumsy but should help beginners...

```suggestion

To try the provisioning service with nRF9151 DK, the device must be claimed using its attestation token, which can be retrieved with ``AT%ATTESTTOKEN`` command.
Please refer to the `nRF Cloud claiming Devices`_ documentation for more information.
Other devices need to follow the instructions from `nRF Cloud Provisioning Service`_ including adding necessary certificates.
Any steps requiring nRF9151 application FW support are handled by the |SM|.

```

These commands are only available when the application is built with the :file:`overlay-nrf-device-provisioning.conf` overlay, which enables the provisioning client and its AT command interface.

.. note::

   The provisioning client requires the host to manage the LTE connection.
   During a provisioning session, the modem must be taken offline to write credentials,
   then brought back online to reconnect to the provisioning server.

>    The device signals these requirements via ``#XNRFPROV: 1`` and ``#XNRFPROV: 2`` notifications.

Reviewer: ```suggestion
   During a provisioning session, the modem must be taken offline to write credentials, then brought back online to reconnect to the provisioning server.
   The device signals these requirements through ``#XNRFPROV: 1`` and ``#XNRFPROV: 2`` notifications.
```

   See the :ref:`unsolicited notifications <xnrfprov_notifications>` section for details.

Trigger provisioning #XNRFPROV
===============================

The ``#XNRFPROV`` command triggers an immediate provisioning attempt.

Set command
-----------

The set command starts a provisioning session immediately, independently of the configured schedule interval.

Syntax
~~~~~~

::

   AT#XNRFPROV

Response syntax
~~~~~~~~~~~~~~~

::

   OK

.. _xnrfprov_notifications:

Unsolicited notifications
~~~~~~~~~~~~~~~~~~~~~~~~~

> One of the following ``#XNRFPROV`` notifications is emitted when the provisioning session completes, an event occurs, or host action is required

Reviewer: ```suggestion
One of the following ``#XNRFPROV`` notifications is emitted when the provisioning session completes, an event occurs, or host action is required:
```

::

   #XNRFPROV: <status>

* The ``<status>`` parameter is an integer.
  It can have the following values:

  * ``0`` - Provisioning successful, or no pending commands on the server.
  * ``1`` - Host action required: deactivate LTE.
    The host must take the modem offline (for example ``AT+CFUN=4``) so that credentials can be written safely.

>     The provisioning client waits until the modem reports offline functional mode before continuing.

Reviewer 3: Does the provision library take care of this? I was kind of expecting to see some code related to this in Serial Modem.

Author: Yes, the provisioning library waits modem state changes.

  * ``2`` - Host action required: activate LTE.
    The host must bring the modem back online (for example ``AT+CFUN=1``) to reconnect to the provisioning server.
    The provisioning client waits until the modem reports LTE registration before continuing.
  * ``-1`` - Provisioning failed.

>     The session will be retried according to the configured interval.

Reviewer 3: There is no interval?

  * ``-2`` - Device not claimed on nRF Cloud.

>     The device must be claimed using its attestation token. Please refer to the `nRF Cloud claiming Devices`_ documentation for more information.

Reviewer: ```suggestion
    The device must be claimed using its attestation token. 
    Refer to the `nRF Cloud claiming Devices`_ documentation for more information.
```

  * ``-3`` - Wrong root CA certificate.
    Provision the correct nRF Cloud root CA certificate.
  * ``-4`` - No valid datetime reference.
    The modem must have a valid time before provisioning can proceed.
  * ``-5`` - Too many commands received from the server.
    Increase ``CONFIG_NRF_PROVISIONING_CBOR_RECORDS``.
  * ``-6`` - Fatal error; the provisioning client encountered an irrecoverable error.

>   * Positive integer - A future scheduled provisioning attempt is pending.

Reviewer 3: Is this still related to interval or does this take place also provisioning attempt that is immediately triggered?

Author: Can be removed

    The value is the number of seconds until the next attempt.

Example
~~~~~~~

> Trigger a provisioning attempt with host-managed LTE

Reviewer 2: Since we don't have interval method, I think this caption can be removed. This example basically covers all we have.

    The provisioning client waits until the modem reports offline functional mode before continuing or times out after ``CONFIG_NRF_PROVISIONING_MODEM_STATE_WAIT_TIMEOUT_SECONDS`` seconds.
  * ``2`` - Host action required: activate LTE.
    The host must bring the modem back online (for example ``AT+CFUN=1``) to reconnect to the provisioning server.
    The provisioning client waits until the modem reports LTE registration before continuing or times out after ``CONFIG_NRF_PROVISIONING_MODEM_STATE_WAIT_TIMEOUT_SECONDS`` seconds.
  * ``-1`` - Provisioning failed.
    The host can retry provisioning by triggering the ``#XNRFPROV`` command again.
  * ``-2`` - Device not claimed on nRF Cloud.
    The device must be claimed using its attestation token.
    Refer to the `nRF Cloud claiming Devices`_ documentation for more information.
  * ``-3`` - Wrong root CA certificate.
    Provision the correct nRF Cloud root CA certificate.
  * ``-4`` - No valid datetime reference.
    The modem must have a valid time before provisioning can proceed.
  * ``-5`` - Too many commands received from the server.

>     Increase ``CONFIG_NRF_PROVISIONING_CBOR_RECORDS``.

Reviewer 2: Could we mention here what is here https://github.com/nrfconnect/ncs-serial-modem/pull/258#discussion_r3129419878. Something like:

```suggestion
    Delete some provisioning commands from the server.
    Or increase ``CONFIG_NRF_PROVISIONING_CBOR_RECORDS``.
```

  * ``-1`` - Provisioning failed.
    The session will be retried according to the configured interval.
  * ``-2`` - Device not claimed on nRF Cloud.
    The device must be claimed using its attestation token. Please refer to the `nRF Cloud claiming Devices`_ documentation for more information.
  * ``-3`` - Wrong root CA certificate.
    Provision the correct nRF Cloud root CA certificate.
  * ``-4`` - No valid datetime reference.
    The modem must have a valid time before provisioning can proceed.
  * ``-5`` - Too many commands received from the server.
    Increase ``CONFIG_NRF_PROVISIONING_CBOR_RECORDS``.
  * ``-6`` - Fatal error; the provisioning client encountered an irrecoverable error.
  * Positive integer - A future scheduled provisioning attempt is pending.
    The value is the number of seconds until the next attempt.

Example
~~~~~~~

Trigger a provisioning attempt
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

::

   AT#XNRFPROV
   OK

   #XNRFPROV: 0

Test command
------------

The test command returns the command syntax.

Syntax
~~~~~~

::

   AT#XNRFPROV=?

Response syntax
~~~~~~~~~~~~~~~

::

   #XNRFPROV

> Set provisioning interval #XNRFPROVINTERVAL

Reviewer 2: Keep this for now. I'll check with management whether we need this.
Developers' initial thought is that this should be handled by the host completely.

Author: It was agreed that this command is not needed.
